url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/gui/component/TextArea.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * This file is part of lanterna (http://code.google.com/p/lanterna/).
+ * 
+ * lanterna is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Copyright (C) 2010-2012 Martin
+ */
+
+package com.googlecode.lanterna.gui.component;
+
+import com.googlecode.lanterna.gui.TextGraphics;
+import com.googlecode.lanterna.gui.Theme;
+import com.googlecode.lanterna.input.Key;
+import com.googlecode.lanterna.terminal.ACS;
+import com.googlecode.lanterna.terminal.TerminalPosition;
+import com.googlecode.lanterna.terminal.TerminalSize;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * This component is designed for displaying large chunks of text. If the text
+ * is larger than the component, it will display scrollbars and letting the
+ * user scroll through the text using the arrow keys.
+ * @author mberglun
+ */
+public class TextArea  extends AbstractInteractableComponent
+{ 
+    private final List<String> lines;
+    private final TerminalSize preferredSize;
+    
+    private TerminalSize lastSize;
+    private TerminalSize maximumSize;
+    private TerminalSize minimumSize;
+    private int longestLine;
+    private int scrollTopIndex;
+    private int scrollLeftIndex;
+
+    public TextArea() {       
+        this("");
+    }
+
+    public TextArea(String text) {       
+        this(new TerminalSize(0, 0), text);
+    }
+    
+    public TextArea(TerminalSize preferredSize, String text) {
+        if(text == null)
+            text = "";
+        
+        this.lines = new ArrayList<String>();
+        this.preferredSize = preferredSize;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;preferredSize&#39; (index: 0 | type: TerminalSize). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;text&#39;. The same parameter type is validated here for example: url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/gui/component/ActionListBox.java:43. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).